File: app/MultiMarket/c/src/data/Manager.c
// Copyright 12-Jul-2019 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "data/Manager.h"
#include "data/ModelMxMn.h"

/* .
ManagerEntry: serial
  model: char *
  params: Darr
===
Manager: serial
  # Default ManagerEntry
  @current: ManagerEntry
  # Map[ManagerEntry]. Keys are nicks.
  entries: Map - ManagerEntry
===
# Format to show parameters in javascript
ManagerFormat: serial
  prefix: char *
  multiplicator: int
  decimals: int
  suffix: char *
===
ManagerEntry2: to
  model: char *
  params: Darr
  param_cf: Arr - ModelMxMn
  param_fmt: Arr - ManagerFormat
===
Manager2: to
  # Default ManagerEntry
  current: ManagerEntry2
  # Map[ManagerEntry2]. Keys are nicks.
  entries: Map - ManagerEntry2
*/

/*--*/

struct Manager_ManagerEntry {
  char *model;
  Darr *params;
};

ManagerEntry *managerEntry_new (char *model, Darr *params) {
  ManagerEntry *this = MALLOC(ManagerEntry);
  this->model = model;
  this->params = params;
  return this;
}

char *managerEntry_model (ManagerEntry *this) {
  return this->model;
}

Darr *managerEntry_params (ManagerEntry *this) {
  return this->params;
}

Js *managerEntry_to_js (ManagerEntry *this) {
  // Arr[Js]
  Arr *js = arr_new();
  arr_push(js, js_ws(this->model));
  arr_push(js, darr_to_js(this->params));
  return js_wa(js);
}

ManagerEntry *managerEntry_from_js (Js *js) {
  // Arr[Js]
  Arr *a = js_ra(js);
  Js **p = (Js **)arr_start(a);
  ManagerEntry *this = MALLOC(ManagerEntry);
  this->model = js_rs(*p++);
  this->params = darr_from_js(*p++);
  return this;
}

struct Manager_Manager {
  ManagerEntry *current;
  Map *entries;
};

Manager *manager_new (ManagerEntry *current, Map *entries) {
  Manager *this = MALLOC(Manager);
  this->current = current;
  this->entries = entries;
  return this;
}

ManagerEntry *manager_current (Manager *this) {
  return this->current;
}

void manager_set_current (Manager *this, ManagerEntry *value) {
  this->current = value;
}

Map *manager_entries (Manager *this) {
  return this->entries;
}

Js *manager_to_js (Manager *this) {
  // Arr[Js]
  Arr *js = arr_new();
  arr_push(js, managerEntry_to_js(this->current));
  arr_push(js, map_to_js(this->entries, (FTO)managerEntry_to_js));
  return js_wa(js);
}

Manager *manager_from_js (Js *js) {
  // Arr[Js]
  Arr *a = js_ra(js);
  Js **p = (Js **)arr_start(a);
  Manager *this = MALLOC(Manager);
  this->current = managerEntry_from_js(*p++);
  this->entries = map_from_js(*p++, (FFROM)managerEntry_from_js);
  return this;
}

struct Manager_ManagerFormat {
  char *prefix;
  int multiplicator;
  int decimals;
  char *suffix;
};

ManagerFormat *managerFormat_new (
  char *prefix,
  int multiplicator,
  int decimals,
  char *suffix
) {
  ManagerFormat *this = MALLOC(ManagerFormat);
  this->prefix = prefix;
  this->multiplicator = multiplicator;
  this->decimals = decimals;
  this->suffix = suffix;
  return this;
}

char *managerFormat_prefix (ManagerFormat *this) {
  return this->prefix;
}

int managerFormat_multiplicator (ManagerFormat *this) {
  return this->multiplicator;
}

int managerFormat_decimals (ManagerFormat *this) {
  return this->decimals;
}

char *managerFormat_suffix (ManagerFormat *this) {
  return this->suffix;
}

Js *managerFormat_to_js (ManagerFormat *this) {
  // Arr[Js]
  Arr *js = arr_new();
  arr_push(js, js_ws(this->prefix));
  arr_push(js, js_wi((int)this->multiplicator));
  arr_push(js, js_wi((int)this->decimals));
  arr_push(js, js_ws(this->suffix));
  return js_wa(js);
}

ManagerFormat *managerFormat_from_js (Js *js) {
  // Arr[Js]
  Arr *a = js_ra(js);
  Js **p = (Js **)arr_start(a);
  ManagerFormat *this = MALLOC(ManagerFormat);
  this->prefix = js_rs(*p++);
  this->multiplicator = js_ri(*p++);
  this->decimals = js_ri(*p++);
  this->suffix = js_rs(*p++);
  return this;
}

struct Manager_ManagerEntry2 {
  char *model;
  Darr *params;
  Arr *param_cf;
  Arr *param_fmt;
};

ManagerEntry2 *managerEntry2_new (
  char *model,
  Darr *params,
  Arr *param_cf,
  Arr *param_fmt
) {
  ManagerEntry2 *this = MALLOC(ManagerEntry2);
  this->model = model;
  this->params = params;
  this->param_cf = param_cf;
  this->param_fmt = param_fmt;
  return this;
}

char *managerEntry2_model (ManagerEntry2 *this) {
  return this->model;
}

Darr *managerEntry2_params (ManagerEntry2 *this) {
  return this->params;
}

Arr *managerEntry2_param_cf (ManagerEntry2 *this) {
  return this->param_cf;
}

Arr *managerEntry2_param_fmt (ManagerEntry2 *this) {
  return this->param_fmt;
}

Js *managerEntry2_to_js (ManagerEntry2 *this) {
  // Arr[Js]
  Arr *js = arr_new();
  arr_push(js, js_ws(this->model));
  arr_push(js, darr_to_js(this->params));
  arr_push(js, arr_to_js(this->param_cf, (FTO)modelMxMn_to_js));
  arr_push(js, arr_to_js(this->param_fmt, (FTO)managerFormat_to_js));
  return js_wa(js);
}

struct Manager_Manager2 {
  ManagerEntry2 *current;
  Map *entries;
};

Manager2 *manager2_new (ManagerEntry2 *current, Map *entries) {
  Manager2 *this = MALLOC(Manager2);
  this->current = current;
  this->entries = entries;
  return this;
}

ManagerEntry2 *manager2_current (Manager2 *this) {
  return this->current;
}

Map *manager2_entries (Manager2 *this) {
  return this->entries;
}

Js *manager2_to_js (Manager2 *this) {
  // Arr[Js]
  Arr *js = arr_new();
  arr_push(js, managerEntry2_to_js(this->current));
  arr_push(js, map_to_js(this->entries, (FTO)managerEntry2_to_js));
  return js_wa(js);
}

/*--*/

File: app/MultiMarket/js/src/data/Manager.js
// Copyright 08-Jun-2019 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

import Dec from "../dmjs/Dec.js";
import It from "../dmjs/It.js";

/// Format to show parameters
export class ManagerFormat {

  /* .
  _rc_ ManagerFormat : from
    prefix: string
    multiplicator: number
    decimals: number
    suffix: string
  */

  /*--*/
  /**
   * @param {string} prefix
   * @param {number} multiplicator
   * @param {number} decimals
   * @param {string} suffix
   */
  constructor (prefix, multiplicator, decimals, suffix) {

    /**
     * @private
     * @type {string}
     */
    this._prefix = prefix;

    /**
     * @private
     * @type {number}
     */
    this._multiplicator = multiplicator;

    /**
     * @private
     * @type {number}
     */
    this._decimals = decimals;

    /**
     * @private
     * @type {string}
     */
    this._suffix = suffix;

  }

  /**  @return {string} */
  get prefix () {
    return this._prefix;
  }

  /**  @return {number} */
  get multiplicator () {
    return this._multiplicator;
  }

  /**  @return {number} */
  get decimals () {
    return this._decimals;
  }

  /**  @return {string} */
  get suffix () {
    return this._suffix;
  }

  /**
   * @param {!Array<?>} serial
   * @return {!ManagerFormat}
   */
  static fromJs (serial) {
    return new ManagerFormat(
      serial[0],
      serial[1],
      serial[2],
      serial[3]
    );
  }
  /*--*/

}

/** Tuple Name-Maximum-Minimum of a parameter. */
export class ModelMxMn {

  /* .
  _rc_ ModelMxMn : from
    name: string
    max : number
    min : number
  */

  /*--*/
  /**
   * @param {string} name
   * @param {number} max
   * @param {number} min
   */
  constructor (name, max, min) {

    /**
     * @private
     * @type {string}
     */
    this._name = name;

    /**
     * @private
     * @type {number}
     */
    this._max = max;

    /**
     * @private
     * @type {number}
     */
    this._min = min;

  }

  /**  @return {string} */
  get name () {
    return this._name;
  }

  /**  @return {number} */
  get max () {
    return this._max;
  }

  /**  @return {number} */
  get min () {
    return this._min;
  }

  /**
   * @param {!Array<?>} serial
   * @return {!ModelMxMn}
   */
  static fromJs (serial) {
    return new ModelMxMn(
      serial[0],
      serial[1],
      serial[2]
    );
  }
  /*--*/

}

/** Manager entry */
export class ManagerEntry {

  /* .
  _rc_ ManagerEntry : from
    model    : string
    params   : !Array<number>
    paramCfs : !Array<!ModelMxMn>
    paramFmts: !Array<!ManagerFormat>
  */

  /*--*/
  /**
   * @param {string} model
   * @param {!Array<number>} params
   * @param {!Array<!ModelMxMn>} paramCfs
   * @param {!Array<!ManagerFormat>} paramFmts
   */
  constructor (model, params, paramCfs, paramFmts) {

    /**
     * @private
     * @type {string}
     */
    this._model = model;

    /**
     * @private
     * @type {!Array<number>}
     */
    this._params = params;

    /**
     * @private
     * @type {!Array<!ModelMxMn>}
     */
    this._paramCfs = paramCfs;

    /**
     * @private
     * @type {!Array<!ManagerFormat>}
     */
    this._paramFmts = paramFmts;

  }

  /**  @return {string} */
  get model () {
    return this._model;
  }

  /**  @return {!Array<number>} */
  get params () {
    return this._params;
  }

  /**  @return {!Array<!ModelMxMn>} */
  get paramCfs () {
    return this._paramCfs;
  }

  /**  @return {!Array<!ManagerFormat>} */
  get paramFmts () {
    return this._paramFmts;
  }

  /**
   * @param {!Array<?>} serial
   * @return {!ManagerEntry}
   */
  static fromJs (serial) {
    return new ManagerEntry(
      serial[0],
      serial[1],
      serial[2].map(e =>
        ModelMxMn.fromJs(e)
      ),
      serial[3].map(e =>
        ManagerFormat.fromJs(e)
      )
    );
  }
  /*--*/

  /**
   * @param {!ManagerEntry} other
   * @return {boolean}
   */
  eqParams (other) {
    const tps = this.params;
    const ops = other.params;
    if (tps.length !== ops.length) {
      return false;
    }
    return It.range(tps.length).every(ix =>
      new Dec(tps[ix], 7).eqValue(new Dec(ops[ix], 7))
    );
  }

}

/** Manager who matches nicks-models*/
export class Manager {

  /* .
  _rc_ Manager : from
    current   : !ManagerEntry
    # Keys are nicks
    entries : !Map<!ManagerEntry>
  */

  /*--*/
  /**
   * @param {!ManagerEntry} current
   * @param {!Map<string, !ManagerEntry>} entries Keys are nicks
   */
  constructor (current, entries) {

    /**
     * @private
     * @type {!ManagerEntry}
     */
    this._current = current;

    /**
     * @private
     * @type {!Map<string, !ManagerEntry>}
     */
    this._entries = entries;

  }

  /**  @return {!ManagerEntry} */
  get current () {
    return this._current;
  }

  /**  @return {!Map<string, !ManagerEntry>} */
  get entries () {
    return this._entries;
  }

  /**
   * @param {!Array<?>} serial
   * @return {!Manager}
   */
  static fromJs (serial) {
    return new Manager(
      ManagerEntry.fromJs(serial[0]),
      new Map(serial[1].map(kv =>
        [kv[0], ManagerEntry.fromJs(kv[1])]
      ))
    );
  }
  /*--*/

}
